docs/tutorials/intro-to-rust.md

```diff
@@ -0,0 +1,418 @@
+
+Below are some snippets from the `lib.rs` file:
+
+```rust
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+#[derive(Serialize, Deserialize)]
+pub struct Counter {
+    val: i8, // i8 is signed. unsigned integers are also available: u8, u16, u32, u64, u128
+}
+
+#[near_bindgen]
+impl Counter {    
+  … 
+```
+
+When writing smart contracts, the pattern is to have a `struct` with an associated `impl` where we write the core logic into functions. It's actually common in Rust to have this pattern elsewhere.
+
+<blockquote class="info">
+  "…most functions will end up being inside impl blocks…"
+  <a href="https://doc.rust-lang.org/std/keyword.fn.html" target="_blank">- Rust docs</a>
+</blockquote>
+
+### Our core logic: the `struct`
+
+We declare our `Counter` and `impl`, defining the functions we'll be invoking on the blockchain.
+
+Above the definitions we see [attributes](https://doc.rust-lang.org/reference/attributes.html) specific to NEAR:
+
+```rust
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+#[derive(Serialize, Deserialize)]
+```
+
+These essentially allow the compilation into WebAssembly to be compatible and optimized for the NEAR blockchain.
+
+We use the context `env` to write logs, as mentioned earlier.
+
+<blockquote class="lesson">
+<code>env</code> can return very useful information including:
+<ul>
+  <li><strong>signer_account_id</strong> - the account id of that signed the original transaction that led to this execution</li>
+  <li><strong>attached_deposit</strong> - if someone sent tokens along with the call</li>
+  <li><strong>account balance</strong> - the balance attached to the given account</li>
+  <li><em>and more…</em></li>
+</ul>
+<p>More info <a href="https://nomicon.io/index.html" target="_blank">available here</a>.</p>
+</blockquote>
+
+### Unit tests
+
+The unit tests begin at:
+
+```rust
+mod tests {
+  …
+}
+```
+
+and continue until the end of the `lib.rs` file. The code here is fairly boilerplate. The custom unit test code comes into play here:
+
+```rust
+let mut contract = Counter{ val: 0 };
+contract.increment();
+// confirm that we received 1 when calling get_num
+println!("Value after increment: {}", contract.get_num());
+assert_eq!(1, contract.get_num());
+```
+
+<blockquote class="lesson">
+<p>Notice the naming convention of variables here. There's a lot of <code>snake_case</code> and variables prefixed with an underscore. Upon building a project, Rust informs you if naming conventions are off. It's actually quite easy to write proper Rust code using the compiler's suggestions.</p>
+</blockquote>
+
+You may add as many tests as you need following the pattern in this file. Similar to unit tests in other languages and frameworks, just add the attribute:
+
+`#[test]`
+
+above the block of code to have it run in the suite of tests.
+
+</hr>
```

Comment:
Did you mean `<hr />`?

Reply:
wow, that's embarrassing

Reply:
<img width="376" alt="Screenshot 2020-04-06 20 46 46" src="https://user-images.githubusercontent.com/1042667/78627996-c2892380-7847-11ea-9158-dc4469d7273b.png">